framework/wcf/Extensibility/Transport/AzureQueueStorageClient/AzureQueueStorageClient/AzureQueueStorageOutputChannel.cs

```diff
@@ -0,0 +1,263 @@
+﻿//----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+//----------------------------------------------------------------
+
+using System;
+using System.Buffers;
+using System.Globalization;
+using System.Net;
+using System.Net.Sockets;
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+using Azure.Storage.Queues;
+using Azure.Storage.Queues.Models;
+
+namespace Microsoft.Samples.AzureQueueStorage
+{
+    /// <summary>
+    /// IOutputChannel implementation for AzureQueueStorage.
+    /// </summary>
+    class AzureQueueStorageOutputChannel : ChannelBase, IOutputChannel
+    {
+        #region member_variables
+        EndpointAddress remoteAddress;
+        Uri via;
+        EndPoint remoteEndPoint;
+        MessageEncoder encoder;
+        AzureQueueStorageChannelFactory parent;
+        QueueClient queueClient;
+        #endregion
```

Comment:
I'm not seeing where you are initializing an instance of this. This is where the plumbing piece comes in. In the constructor, you should get the via address from the factory, and then use that for constructing the QueueClient. Remember as we spoke that you need to handle either an https:// address or a net.aqs:// address and convert the net.aqs address to the https:// address to pass to the queue client. The Via address used in the To header needs to be a net.aws:// address so the service will route it consistently so you will need to be able to convert both ways. You also haven't provided a way to provide credentials to the QueueClient. I'll follow up with another comment about that after I've done some research.

Reply:
I had a look at the QueueClient api for providing authentication and it's a little bit chaotic. They don't have a single mechanism to provide the authentication credentials as you can provide a SAS token as a uril parameter, a connection string, use an AzureSasCredential object, a TokenCredential, or a StorageSharedKeyCredential. We can't possible provide a clean api to enable all of those options. The most flexible seems to be using a connection string as you can provide a SaS credential or a shared key, It's intended that you can store this in Azure Key Value, and it fits into the .net configuration api story where you can retrieve a string from config with multiple ways to store it.  
There are 2 limitations. The constructor for QueueClient takes the uri from the connection string, so we now have 2 places where the uri can come from. This isn't too big of a problem as WCF has a concept of To and Via. Via is the url your raw transport connects to, and To is used in the To header of the soap message, and we do have the concept that these can be different. So the Via property will get populated from the QueueClient as I believe you can request from it what the url you are connecting to is. And the url provided to the endpoint in the channel factory constructor will be used to populate the To header of the soap message.  
The second limitation is that this is insufficient for all scenarios. We can use a mechanism I created for Http where you want to modify, replace, or wrap the HttpClientHandler. We support adding via an endpoint behavior BindingParameters. For the Http transport, you can add a `Func<HttpClientHandler, HttpMessageHandler>` to the binding parameters and WCF will construct its own HttpClientHandler and configure it, then pass it to this func, and use the returned HttpMessageHandler when creating the HttpClient. We can do a similar thing here where we say best practice and first class api support is to use a connection string. If you need one of the other mechanisms, add a delegate to the binding parameters which we will call and allows them to provide a custom QueueClient instance. We might need to check a few things on the returned client, eg the message encoding value is what we need (unless it can be modified afterwards?). I think this is the best compromise, each public api to use which matches best practices, and a slightly awkward (we can provide a helper method to take the boiler plate out, which I can help you write) alternative way to do things differently.

Reply:
[This is the code in the HttpChannelFactory](https://github.com/dotnet/wcf/blob/2bfdb6425b50fd5f22353c887dd29fc3ef906c37/src/System.ServiceModel.Http/src/System/ServiceModel/Channels/HttpChannelFactory.cs#L131) where we grab the Func from the binding parameters. As you can see, it's relatively uncomplicated to retrieve. Then in the implementation, after you construct you own instance of QueueClient, you check if this Func is not null, and call it, replacing your own QueueClient with what was returned.